### PR TITLE
Dynamic model for `PositioningInformationView`

### PR DESCRIPTION
**Pull request by mohsenD98**
### Description:

This PR introduces enhancements to the PositioningInformationView.

---

### Current State:

The current PositioningInformationView displays static `Text { }` elements in QML that vary in visibility based on the type of receiver (internal or external NMEA). The goal is to support displaying any details a receiver device can provide. 

**Internal devices** feature information such as:

    - Latitude/Longitude (x/y)
    - Altitude
    - Speed
    - Horizontal Accuracy
    - Vertical Accuracy

External **NMEA** receivers provide these, plus additional attributes like:

    - PDOP
    - HDOP
    - VDOP
    - Validity
    - Fix Status
    - Quality

---

### New State:

We are replacing the static `Text { }` components with a dynamic `GridView { }` backed by a custom model, `PositioningInformationModel`. This model encompasses six core attributes that are common across all receivers: `Latitude, Longitude, Altitude, Speed, Vertical Accuracy, and Horizontal Accuracy`. Additionally, there's a `deviceDetails` variable of type `QList<QPair<QString, QVariant>>`.

The `deviceDetails` list will hold other parameters specific to each receiver. For instance, in the `NmeaGnssReceiver`, this implementation would look like:
``` C++
QList<QPair<QString, QVariant>> NmeaGnssReceiver::details()
{
  QList<QPair<QString, QVariant>> dataList;

  dataList.append(qMakePair("PDOP", QLocale::system().toString(mCurrentNmeaGnssPositionInformation.pdop(), 'f', 1)));
  dataList.append(qMakePair("HDOP", QLocale::system().toString(mCurrentNmeaGnssPositionInformation.hdop(), 'f', 1)));
  dataList.append(qMakePair("VDOP", QLocale::system().toString(mCurrentNmeaGnssPositionInformation.vdop(), 'f', 1)));
  dataList.append(qMakePair("Valid", mCurrentNmeaGnssPositionInformation.isValid() ? "True" : "False"));
  dataList.append(qMakePair("Fix", mCurrentNmeaGnssPositionInformation.fixStatusDescription()));
  dataList.append(qMakePair("Quality", mCurrentNmeaGnssPositionInformation.qualityDescription()));

  return dataList;
}
```
With this update, when a new receiver type is introduced, the only requirement is to implement the details function, allowing the UI to update seamlessly and automatically.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 745c49fb81b86cae2bc7383a78e961fb65e96890: [**arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5604-8c63c248c8e1ff87bf70b93e93f45f152142c7c7-android-arm64.apk)
<details>
<summary>Other architectures</summary>

- [arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5604-8c63c248c8e1ff87bf70b93e93f45f152142c7c7-android-arm-neon.apk)
- [x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5604-8c63c248c8e1ff87bf70b93e93f45f152142c7c7-android-x64.apk)
- [x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5604-8c63c248c8e1ff87bf70b93e93f45f152142c7c7-android-x86.apk)
- [**all access arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5604-8c63c248c8e1ff87bf70b93e93f45f152142c7c7-android-arm64.apk)
- [all access arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5604-8c63c248c8e1ff87bf70b93e93f45f152142c7c7-android-arm-neon.apk)
- [all access x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5604-8c63c248c8e1ff87bf70b93e93f45f152142c7c7-android-x64.apk)
- [all access x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5604-8c63c248c8e1ff87bf70b93e93f45f152142c7c7-android-x86.apk)
</details>

File: src/qml/InformationDrawer.qml [hunk 1]
```diff
@@ -65,10 +65,10 @@ Item {
       PositioningInformationView {
         id: positioningInformationView
         width: parent.width
-        height: contentHeight
+        height: Math.min(contentHeight, mainWindow.height / 3)
         visible: positioningInformationViewEnabled
         positionSource: controller.positionSource
-        antennaHeight: positioningSettings.antennaHeightActivated ? positioningSettings.antennaHeight : NaN
+        antennaHeight: positioningSettings.antennaHeightActivated ? positioningSettings.antennaHeight : -1
```
Comment (nirvn) [inline src/qml/InformationDrawer.qml:71]: -1 is valid. Negative antenna  height is allowed.

File: src/qml/PositioningInformationView.qml [hunk 2]
```diff
@@ -8,372 +8,59 @@ import Theme
...
+  property double cellHeight: 26
+  property double cellPadding: 6
+  property real contentHeight: grid.count / grid.numberOfColumns * cellHeight
 
   color: Theme.mainBackgroundColorSemiOpaque
+  anchors.margins: 20
+  width: parent.width
 
-  Grid {
+  GridView {
```
Comment (nirvn) [inline src/qml/PositioningInformationView.qml:25]: Sweet -- did you check that the details ordering remain the same as what we had before, inc. for both 2 cells per row vs. 4 cells per row?

File: src/qml/InformationDrawer.qml (hunk 1 above)
Comment (mohsenD98) [reply src/qml/InformationDrawer.qml:71]: Ok, so previously it was setting `NaN` and for altitude :

```QML
if (!isNaN(parseFloat(antennaHeight))) {
  details.push('ant.');
}
``` 

any suggestion ?

File: src/qml/PositioningInformationView.qml (hunk 2 above)
Comment (mohsenD98) [reply src/qml/PositioningInformationView.qml:25]: yes.

File: src/qml/InformationDrawer.qml (hunk 1 above)
Comment (mohsenD98) [reply src/qml/InformationDrawer.qml:71]: i tried to assign `NaN` to a Q_PROPERTY with type `double`, it simply ignored it and value did not changed.

File: src/core/positioning/positioninginformationmodel.cpp [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+  getCoordinateLabels( coord1Label, coord2Label, coordinatesIsXY, coordinatesIsGeographic );
+  getCoordinateValues( coord1Value, coord2Value, coordinates, coordinatesIsXY, coordinatesIsGeographic );
+
+  const QString altitude = getAltitude( distanceUnitFactor, distanceUnitAbbreviation );
+  const QString speed = getSpeed();
+  const QString hAccuracy = getHorizontalAccuracy( distanceUnitFactor, distanceUnitAbbreviation );
+  const QString vAccuracy = getVerticalAccuracy( distanceUnitFactor, distanceUnitAbbreviation );
+
+  updateInfo( coord1Label, coord1Value );
+  updateInfo( coord2Label, coord2Value );
```
[the remark this answers is not in the document]
Comment (mohsenD98) [reply src/core/positioning/positioninginformationmodel.cpp:40]: is this ok? 
```C++
if ( (rowItem->data( NameRole ).toString() == "X" && name == "Lon") ||
  (rowItem->data( NameRole ).toString() == "Y" && name == "Lat") ||
  (rowItem->data( NameRole ).toString() == "Lon" && name == "X") ||
  (rowItem->data( NameRole ).toString() == "Lat" && name == "Y") )
  {
      rowItem->setData( name, NameRole );
      rowItem->setData( value.toString(), ValueRole );
      return;
  }
```

File: src/core/positioning/positioninginformationmodel.cpp [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+
+  const QString altitude = getAltitude( distanceUnitFactor, distanceUnitAbbreviation );
+  const QString speed = getSpeed();
+  const QString hAccuracy = getHorizontalAccuracy( distanceUnitFactor, distanceUnitAbbreviation );
+  const QString vAccuracy = getVerticalAccuracy( distanceUnitFactor, distanceUnitAbbreviation );
+
+  updateInfo( "Altitude", altitude );
+  updateInfo( "Speed", speed );
+  updateInfo( "H. Accuracy", hAccuracy );
+  updateInfo( "V. Accuracy", vAccuracy );
```
Comment (nirvn) [inline src/core/positioning/positioninginformationmodel.cpp:36]: tr() all of this please.

File: src/core/positioning/positioninginformationmodel.cpp [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+    {
+      details.push_back( "grid" );
+    }
+    else if ( positioningSource()->elevationCorrectionMode() == Positioning::ElevationCorrectionMode::OrthometricFromDevice )
+    {
+      details.push_back( "ortho." );
+    }
+    if ( antennaHeight() != -1 )
+    {
+      details.push_back( "ant." );
```
Comment (nirvn) [inline src/core/positioning/positioninginformationmodel.cpp:106]: tr() the strings here too.

File: src/core/positioning/positioninginformationmodel.cpp [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+      rowItem->setData( value.toString(), ValueRole );
+      return;
+    }
+
+    if ( ( rowItem->data( NameRole ).toString() == "X" && name == "Lon" ) || ( rowItem->data( NameRole ).toString() == "Y" && name == "Lat" ) || ( rowItem->data( NameRole ).toString() == "Lon" && name == "X" ) || ( rowItem->data( NameRole ).toString() == "Lat" && name == "Y" ) )
+    {
+      rowItem->setData( name, NameRole );
+      rowItem->setData( value.toString(), ValueRole );
+      return;
+    }
```
Comment (nirvn) [inline src/core/positioning/positioninginformationmodel.cpp:149]: This isn't needed anymore right?

**Review by nirvn — APPROVED** [review]
Alright, let's merge this.

File: src/core/positioning/positioninginformationmodel.cpp [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+}
+
+void PositioningInformationModel::refreshData()
+{
+  if ( !mPositioningSource )
+  {
+    return;
+  }
+
+  const QgsPoint coordinates = GeometryUtils::reprojectPoint( positioningSource()->sourcePosition(), CoordinateReferenceSystemUtils::wgs84Crs(), coordinateDisplayCrs() );
```
Comment (nirvn) [inline src/core/positioning/positioninginformationmodel.cpp:21]:
```suggestion
```

**Pull request merged**